### PR TITLE
JSDK-2469: run integration tests inside container

### PR DESCRIPTION
**Pull request by makarandp0**
This change runs integration tests inside docker containers. It also generates structured report files for the tests and asks travis to upload to S3. The S3 upload does not work for pull request builds. But works for scheduled builds and kicked off build.

one such build in Progress: https://travis-ci.org/twilio/twilio-video.js/builds/571064856 
 
1. DockerFile and Docker-compose.yml file
2. .travis.yml - Update to to enable docker, and execute tests in container
3.  karma/makeconf.js - 
Changes to update browser launch options when running in docker.
Changes to add structured reporting for test results.
Additional option process.env.DEBUG that can help debug by preserving line numbers, and keeping browser open after failures for further inspection. 
4. scripts/karma.js
Changes to reliably make `FILE:foo npm run test:integration` work 

This change folds in changes from https://github.com/twilio/twilio-video.js/pull/709 - Generate and upload build results to S3

Some results from above job 

Firefox Disconnect failures: [1](https://s3.amazonaws.com/twilio-video-logs/twilio/twilio-video.js/1848/1848.8/testresults-BROWSER-firefox-BVER-stable-TOPOLOGY-group-0.html), [2](https://s3.amazonaws.com/twilio-video-logs/twilio/twilio-video.js/1848/1848.7/testresults-BROWSER-firefox-BVER-stable-TOPOLOGY-peer-to-peer-0.html)

Successes: [1](https://s3.amazonaws.com/twilio-video-logs/twilio/twilio-video.js/1848/1848.1/testresults-BROWSER-chrome-BVER-stable-TOPOLOGY-peer-to-peer-2.html), [2](https://s3.amazonaws.com/twilio-video-logs/twilio/twilio-video.js/1848/1848.1/testresults-BROWSER-chrome-BVER-stable-TOPOLOGY-peer-to-peer-1.html), [3](https://s3.amazonaws.com/twilio-video-logs/twilio/twilio-video.js/1848/1848.2/testresults-BROWSER-chrome-BVER-stable-TOPOLOGY-group-1.html), [4](https://s3.amazonaws.com/twilio-video-logs/twilio/twilio-video.js/1848/1848.2/testresults-BROWSER-chrome-BVER-stable-TOPOLOGY-group-2.html)

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -80,31 +83,16 @@ matrix:
...
+  - docker-compose run test npm run test:integration
 
 notifications:
   slack:
     on_pull_requests: false
     secure: jxmuMo2n/OCaUg6DMZ/fahcxoCmkyOt4eD5pWgIqinxL0INyByZhw3EyAItueuXo4+7oh2KZj78ryf2bChSMsGGXXF25Vdy154bGsFXS7xPH/ars7yTo2Cjry8j3ki0ve7aqH3GjMlBVuJRA3ps2w/bJjw9oq4JDiibDRsDPawcS3Xda0tHqc022lJer9jHaF+zBrjobo2xAZ8dFhMmx50wsY5tALq5xoiiAAfmARm8ZIZu6hZpwdbpC6pM9SBd1RNMxnzQrlFPqA25FYPZSaD4/5wjKDEHTO+dfOEW79sgZUuohE/AC+zKd3wiYxxDaxrkPkzijmYs39LyDqm5Vr6eS9OSrZptgGni6G/2J/dBW52MLlxCDB+SqpbaPuLVKsvPdVAwydMTYo02fJpxa3purUdNM9GFPZCZC4KGvht2cSvM1txUNFkpkAp8RCxyUGpVw58S4gnAp2sYmRR22i42nZJunR7myXiMV+KZcaS7BJ2Mf6Wo1YwSnHSP4xwuaBPBY0L86w/gyd9tABpEIamOwKkFiLmOKr1l8aCcuGrEHnBB2L5EpiNfxl93DjrApzwR3FqCeSzfwHgaZ6EIGtRqPPXUJgM4DtYt9udGO4c9fLr6QbDgyS4swn3uUiUmeI3i9UUg92jvv7OCYoe80DfsF5RVjZxXcQM+WMMsJ7os=
+
+addons:
+  artifacts: true
+
```
Comment (manjeshbhargav) [inline .travis.yml]: @makarandp0 Nitpick: remove extra line.

File: karma/makeconf.js [hunk 2]
```diff
@@ -1,26 +1,32 @@
...
+  let files = [];
+  if (config.files && config.files.length) {
+    files = config.files;
+  } else if (defaultFile) {
+    files = [defaultFile];
+  }
+  return files;
+}
+
+let testRun = 0;
```
Comment (manjeshbhargav) [inline karma/makeconf.js]: @makarandp0 Nitpick: add newline between variable and function.

File: .env [hunk 3]
```diff
@@ -0,0 +1,6 @@
+BROWSER=chrome
```
Comment (manjeshbhargav) [inline .env]: @makarandp0 Is this file necessary to be committed, since you're adding it to `.gitignore`?

File: docker/test/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+# It's a good idea to use dumb-init to help prevent zombie chrome processes.
+ADD https://github.com/Yelp/dumb-init/releases/download/v1.2.0/dumb-init_1.2.0_amd64 /usr/local/bin/dumb-init
+RUN chmod +x /usr/local/bin/dumb-init
+
+ADD ./ /opt/app
+WORKDIR /opt/app
+RUN npm install
+
+# Build Args
+ARG USE_BROWSER='firefox'
```
Comment (manjeshbhargav) [inline docker/test/Dockerfile]: @makarandp0 Why are we hard-coding this value?

File: docker/test/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ARG USE_BVER='unstable'
```
Comment (manjeshbhargav) [inline docker/test/Dockerfile]: @makarandp0 Why are we hard-coding this value?

File: docker/test/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+ENV ENABLE_REST_API_TESTS=1
```
Comment (manjeshbhargav) [inline docker/test/Dockerfile]: @makarandp0 Why are we hard-coding this value? This should be enabled only for non peer-to-peer rooms.

File: karma/makeconf.js [hunk 7]
```diff
@@ -29,12 +35,26 @@ function makeConf(defaultFile, browserNoActivityTimeout, requires) {
...
       }
+    } else if (isDocker) {
+      browsers = ['ChromeInDocker', 'FirefoxInDocker'];
     } else if (process.platform === 'darwin') {
       browsers = ['ChromeWebRTC', 'ElectronWebRTC', 'FirefoxWebRTC', 'Safari'];
     } else {
       browsers = ['ChromeWebRTC', 'FirefoxWebRTC'];
     }
 
+    const strTestRun = (testRun++).toString();
```
Comment (manjeshbhargav) [inline karma/makeconf.js]: @makarandp0 Do we need to append the test run number? I think the combination of browser/bver/topology would be unique enough no? Also, I think we should enable test reporting only for docker container. I don't want to push reports when I run integration tests locally.

File: karma/makeconf.js [hunk 8]
```diff
@@ -46,19 +66,48 @@ function makeConf(defaultFile, browserNoActivityTimeout, requires) {
...
+        debug: !!process.env.DEBUG,
         transform: [
           'envify'
         ]
       },
-      reporters: ['spec'],
+      reporters: ['spec', 'junit', 'html'],
+      htmlReporter: { // configuration for karma-htmlfile-reporter
+        outputFile: htmlReport,
+        pageTitle: 'Twilio-Video Integration Tests',
```
Comment (manjeshbhargav) [inline karma/makeconf.js]: @makarandp0 Nitpick: `s/Twilio-Video/twilio-video.js`

File: karma/makeconf.js (hunk 7 above)
Comment (manjeshbhargav) [reply karma/makeconf.js]: @makarandp0 Maybe instead of `testRun`, we can use the Travis build number if we have access to it?

File: scripts/karma.js [hunk 9]
```diff
@@ -10,6 +10,9 @@ const configFile = join(__dirname, '..', process.argv[2]);
 const integrationTests = join(__dirname, '..', 'test', 'integration', 'spec');
 
 function getTestPaths(path) {
+  if (process.env.FILE) {
```
Comment (manjeshbhargav) [inline scripts/karma.js]: @makarandp0 I'm not entirely clear why this is added?

File: .env (hunk 3 above)
Comment (makarandp0) [reply .env]: This file is optional - but  helps with local runs, so I am committing it as a recommended template.  I added it gitignore so that git won't track any local changes to this file.

File: docker/test/Dockerfile (hunk 4 above)
Comment (makarandp0) [reply docker/test/Dockerfile]: it's default value for args and not strictly necessary. It will be overwritten by the parameter passed at build time - and would then be baked into the docker image.

The idea is each docker image is built for specific combination of BROWSER+version 
the BROWSER and version to use is determined by the parameter. If you forget to pass the browser+version, it would default to these values. 

Docker-compose passes these arguments at build time by looking at your env variables. 
https://github.com/twilio/twilio-video.js/blob/JSDK-2469_dockerized_tests/docker-compose.yml#L13

but somebody were to use this DockerFile directly (manually) then its good to have some defaults (instead of building image w/o browser)

File: docker/test/Dockerfile (hunk 5 above)
Comment (makarandp0) [reply docker/test/Dockerfile]: same as above

File: docker/test/Dockerfile (hunk 6 above)
Comment (makarandp0) [reply docker/test/Dockerfile]: I didn't find a nice way to conditionally declare this parameter...but we are okay here, because the test themselves skip if the build is not using p2p rooms. https://github.com/twilio/twilio-video.js/blob/JSDK-2469_dockerized_tests/test/integration/spec/rest.js#L36

File: karma/makeconf.js (hunk 7 above)
Comment (makarandp0) [reply karma/makeconf.js]: unfortunately, the combination of browser+bver+topology is not unique, as we have some custom logic to split tests per file to workaround some memory leaks (https://github.com/twilio/twilio-video.js/blob/JSDK-2469_dockerized_tests/scripts/karma.js#L32). 

I considered using date/time stamp instead of the testRun, but decided against it, as I found simple suffixes makes more deterministic and more readable 

Travis build number does not help either, as they are part of the same travis build (btw build is already encoded in the uri as seen in example files below) 

https://s3.amazonaws.com/twilio-video-logs/twilio/twilio-video.js/1848/1848.2/testresults-BROWSER-chrome-BVER-stable-TOPOLOGY-group-2.html
https://s3.amazonaws.com/twilio-video-logs/twilio/twilio-video.js/1848/1848.2/testresults-BROWSER-chrome-BVER-stable-TOPOLOGY-group-1.html

File: scripts/karma.js (hunk 9 above)
Comment (makarandp0) [reply scripts/karma.js]: We had a similar check earlier in https://github.com/twilio/twilio-video.js/blob/master/karma/makeconf.js#L9 which did not quite work for some files. Moving it at this level makes it works better.

File: karma/makeconf.js (hunk 7 above)
Comment (makarandp0) [reply karma/makeconf.js]: about pushing reports - the reports are generated all the time, but pushed to S3 by Travis only for CI runs. They don't get pushed for local runs.

File: .env (hunk 3 above)
Comment (manjeshbhargav) [reply .env]: @makarandp0 Then I think we should rename it to `.env.template`.

File: .gitignore [hunk 10]
```diff
@@ -13,5 +13,6 @@ npm-debug.log
 package-lock.json
 test.json
 yarn.lock
+.env
```
Comment (syerrapragada) [inline .gitignore]: Nitpick: .env -> .env.template

File: docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,22 @@
+version: '3'
+# to run test in docer:
```
Comment (syerrapragada) [inline docker-compose.yml]: Nitpick: docer -> docker

File: docker/test/Dockerfile (hunk 4 above)
Comment (syerrapragada) [reply docker/test/Dockerfile]: Default in Karma config is chrome. Should we default this to chrome as well?

File: karma/makeconf.js [hunk 12]
```diff
@@ -71,6 +121,19 @@ function makeConf(defaultFile, browserNoActivityTimeout, requires) {
...
+          base: 'Firefox',
+          flags: [
+            '-headless',
+          ],
+          prefs: {
+            'media.gstreamer.enabled': false,
+            'media.navigator.permission.disabled': true,
+            'media.navigator.streams.fake': true,
+            'media.autoplay.enabled.user-gestures-needed': false,
+            'media.block-autoplay-until-in-foreground': false
```
Comment (syerrapragada) [inline karma/makeconf.js]: You may want to consider setting following 2 values as well:
'media.getusermedia.insecure.enabled'
'media.devices.insecure.enabled'

Future builds of FF does not support connecting to insecure URLs, I assume that including localhost.

File: docker/test/Dockerfile (hunk 4 above)
Comment (makarandp0) [reply docker/test/Dockerfile]: Its best to setup env variables/.env files. I intentionally kept defaults to atypical  ff/unstable so that the developer does not forget them. ( tests are more likely to fail in firefox/unstable :)

File: karma/makeconf.js (hunk 12 above)
Comment (makarandp0) [reply karma/makeconf.js]: Thanks, I will experiment with those and add them if they work well.

File: .gitignore (hunk 10 above)
Comment (makarandp0) [reply .gitignore]: @syerrapragada , since we expect everybody to have their version of `.env` - .env needs to go in `.gitignore` to not include track any changes to it. We do want to track any changes to .env.template!
